### PR TITLE
add restoredb command

### LINKED ISSUE #32: restoredb
restore an odoo backup and associated files extracted by backupdb. As for Odoo we must be able to specify if it's a copy or a move...

### PR DESCRIPTION
**Pull request by lmignon**
fixes #32 
includes #58

### CODE REVIEW

**Comment by Olivier-LAURENT** [comment]
Tested by file (not by folder) with success. Thanks, 👍

**Comment by GSLabIt** [comment]
@lmignon is this ready to merge?

**Comment by lmignon** [comment]
@GSLabIt IMO yes it is. I'll resolve the conflicts into the next days

**Comment by lmignon** [comment]
@sbidoul @GSLabIt rebased and ready to merge

**Comment by sbidoul** [comment]
@lmignon don't we need to finish #58 first?

**Comment by lmignon** [comment]
> @lmignon don't we need to finish #58 first?

indeed

**Comment by lmignon** [comment]
@sbidoul  #58 finished

**Comment by GSLabIt** [comment]
@sbidoul any news on this? ready to merge? thx

**Comment by GSLabIt** [comment]
hi all, any chance to get this merged? thx

**Comment by lmignon** [comment]
@GSLabIt thank you for the reminder. This PR will be merged into the next weeks... we're a bit overloaded at the moment which leaves us little time to finalize the various reviews in progress.

**Comment by GSLabIt** [comment]
@lmignon thx for the feedback, looking forward for the merge.

**Comment by rousseldenis** [comment]
@lmignon @sbidoul We need to ask user or to add an option to erase filestore directory. As if it exists, the command raises an error.

**Comment by rousseldenis** [comment]
> @lmignon @sbidoul We need to ask user or to add an option to erase filestore directory. As if it exists, the command raises an error.

But maybe in a ROADMAP so we can move forward on this

**Comment by voronind** [comment]
Raises error in Odoo 10.0
```console
Traceback (most recent call last):
  File "/home/voronin/.local/share/virtualenvs/sintez_addons-7QRHjYmJ/lib/python2.7/site-packages/click_odoo/env_options.py", line 218, in _invoke
    return self.org_invoke(ctx)
  File "/home/voronin/.local/share/virtualenvs/sintez_addons-7QRHjYmJ/lib/python2.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/voronin/.local/share/virtualenvs/sintez_addons-7QRHjYmJ/lib/python2.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/voronin/.local/share/virtualenvs/sintez_addons-7QRHjYmJ/src/click-odoo-contrib/click_odoo_contrib/restoredb.py", line 109, in main
    _restore_from_file(dbname, backup, copy)
  File "/home/voronin/.local/share/virtualenvs/sintez_addons-7QRHjYmJ/src/click-odoo-contrib/click_odoo_contrib/restoredb.py", line 54, in _restore_from_file
    odoo.service.db.restore_db(dbname, backup_file, copy)
  File "/home/voronin/.local/share/virtualenvs/sintez_addons-7QRHjYmJ/lib/python2.7/site-packages/odoo/service/db.py", line 262, in restore_db
    if odoo.tools.exec_pg_command(pg_cmd, *pg_args):
  File "/home/voronin/.local/share/virtualenvs/sintez_addons-7QRHjYmJ/lib/python2.7/site-packages/odoo/tools/misc.py", line 123, in exec_pg_command
    rc = subprocess.call(args2, env=env, stdout=dn, stderr=subprocess.STDOUT)
  File "/usr/lib/python2.7/subprocess.py", line 172, in call
    return Popen(*popenargs, **kwargs).wait()
  File "/usr/lib/python2.7/subprocess.py", line 394, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1047, in _execute_child
    raise child_exception
TypeError: coercing to Unicode: need string or buffer, file found```

**Comment by lmignon** [comment]
> Raises error in Odoo 10.0
>

Thank you @voronind for the report. Unfortunately, it is very difficult to start investigating a potential problem with only a stack-trace :smirk: Can you provide more information on how you use the script. On my side, I've been using it without any problem for months also on Odoo 10.

Regards

**Comment by voronind** [comment]
I try restore Odoo Postgres Dump (not Zip with files). And AFAIK
`odoo.tools.exec_pg_command` get file object instead of string with file
path.

чт, 27 авг. 2020 г. в 10:41, Laurent Mignon (ACSONE) <
notifications@github.com>:

File: click_odoo_contrib/restoredb.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                with env.cr.savepoint():
+                    env.cr.execute("CREATE EXTENSION unaccent")
+            except psycopg2.Error:
+                pass
+    odoo.sql_db.close_db(dbname)
+
+
+def _restore_from_file(dbname, backup, copy=True):
+    with db_management_enabled(), open(backup, "rb") as backup_file:
+        odoo.service.db.restore_db(dbname, backup_file, copy)
```
Comment (voronind) [inline click_odoo_contrib/restoredb.py:54]: Odoo 10 wants file path here? Not file object.

**Comment by lmignon** [comment]
> I try restore Odoo Postgres Dump (not Zip with files). And AFAIK `odoo.tools.exec_pg_command` get file object instead of string with file path.

@voronind there is probably a misunderstanding in the way of using the script. The input must be a path to an Odoo backup zip file or a path to a backup directory created by the click-odoo-backupdb command.

**Comment by voronind** [comment]
> > I try restore Odoo Postgres Dump (not Zip with files). And AFAIK `odoo.tools.exec_pg_command` get file object instead of string with file path.
> 
> @voronind there is probably a misunderstanding in the way of using the script. The input must be a path to an Odoo backup zip file or a path to a backup directory created by the click-odoo-backupdb command.

Then I request feature. Restore Odoo backup __dump__ file.

**Comment by lmignon** [comment]
> Then I request feature. Restore Odoo backup dump file.

@voronind You're welcome to contribute to this PR

**Comment by elvise** [comment]
Hi guys what's the status of this ?

**Comment by sbidoul** [comment]
Hi folks, sorry for the delay with this PR. I hope come back to it in the next month or so, unless @lmignon beats me to it :)

**Comment by sbidoul** [comment]
Rebased. If green tomorrow I'll merge and release.

**Comment by sbidoul** [comment]
@rousseldenis @voronind I'm going to release this as is, I think it is good enough. If you have identified issues or missing feature, can you please open separate issues to discuss them ?

**Comment by GSLabIt** [comment]
@sbidoul pip package will be released as well?

**Comment by sbidoul** [comment]
The release is on it's way to PyPI, as soon as travis catches up: https://travis-ci.com/github/acsone/click-odoo-contrib/builds/187542839

**Pull request merged**